client/asset/btc/btc.go

```diff
@@ -1136,132 +1153,320 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
+			// would be captured when a new tip is reported.
+			btc.findRedemptionError(contractOp, "mempool", err, false)
+			return
 		}
-		for _, txHash := range txs {
-			tx, err := btc.node.GetRawTransactionVerbose(txHash)
+		for i, input := range tx.Vin {
+			secret, err := btc.tryExtractContractSecret(input, contractOp, req.contractHash)
 			if err != nil {
-				return nil, fmt.Errorf("error encountered retreving mempool transaction %s: %v", txHash, err)
-			}
-			for i := range tx.Vin {
-				if tx.Vin[i].Txid == txid && tx.Vin[i].Vout == vout {
-					// Found it. Extract the key.
-					vin := tx.Vin[i]
-					sigScript, err := hex.DecodeString(vin.ScriptSig.Hex)
-					if err != nil {
-						return nil, fmt.Errorf("error decoding scriptSig '%s': %v", vin.ScriptSig.Hex, err)
-					}
-					key, err := dexbtc.FindKeyPush(sigScript, contractHash, btc.chainParams)
-					if err != nil {
-						return nil, fmt.Errorf("found spending input at %s:%d, but error extracting key from '%s': %v",
-							txHash, i, vin.ScriptSig.Hex, err)
-					}
-					return key, nil
-				}
+				// Spending tx found but secret could not be extracted.
+				// Remove from the findRedemptionQueue.
+				btc.findRedemptionError(contractOp, "mempool", err, true)
+			} else if secret != nil {
+				btc.foundRedemption("mempool", contractOp, tx.Txid, i, secret)
 			}
 		}
-		return nil, fmt.Errorf("no key found")
-	}
-	// Since it's not a mempool transaction, start scanning blocks at the height
-	// of the swap.
-	blockHash, err := chainhash.NewHashFromStr(tx.BlockHash)
-	if err != nil {
-		return nil, fmt.Errorf("error decoding block hash: %v", err)
 	}
-	contractBlock := *blockHash
-	var contractHash []byte
-	blockHeight := tx.BlockIndex
-	for {
-		select {
-		case <-ctx.Done():
-			return nil, fmt.Errorf("redemption search canceled")
-		default:
-		}
-		block, err := btc.getVerboseBlockTxs(blockHash.String())
-		if err != nil {
-			return nil, fmt.Errorf("error fetching verbose block '%s': %v", blockHash, err)
+}
+
+// findRedemptionInBlockRange attempts to find spending info for the specified
+// contract by searching every input of all txs in the provided block range.
+// If found, the contract's secret  or any error encountered while trying to
+// parse the contract's secret is returned to the caller via the registered
+// result chan. The contract is only retained in the findRedemptionQueue if
+// no tx was found in the specified block range that redeems it.
+func (btc *ExchangeWallet) findRedemptionInBlockRange(startBlock, endBlock *block, contractOp outPoint) {
+	if endBlock == nil {
+		btc.log.Debugf("finding redemption for contract %s in block %d - best block",
+			contractOp.String(), startBlock.height)
+	} else {
+		btc.log.Debugf("finding redemption for contract %s in blocks %d - %d",
+			contractOp.String(), startBlock.height, endBlock.height)
+	}
+
+	var lastScannedBlock uint64
+	searchTxsForRedemption := func(blkHeight uint64, txs []btcjson.TxRawResult) bool {
```

Comment:
Read somewhere that slices are usually passed by reference, so `txs` here should not be a copy of the data passed in the call to `searchTxsForRedemption`; modifying `txs` here would modify the original data.

I think where there'd be concern for avoidable copying is in the loop that ranges `txs`. I'll improve that.

Reply:
I did not realize data was being modified.

Reply:
It's not; my point was, if modifying the slice here would modify the original slice, then it's not a copy. It's accessing the individual txs in the loop that resulted in a copy. So I changed that to use a pointer instead. I used `[]btcjson.TxRawResult` instead of `[]*btcjson.TxRawResult` because the former is what's returned from `getVerboseBlockTxs`. Instead of looping through the slice to convert each item to a pointer before passing here, I figured it may be faster to convert to pointer here inside the loop that uses each item.